import os
import pandas as pd

locations = list()

with open('/media/sf_Shared/bespoke_mutations_and_primers.txt','r') as mutations_and_primers:
	for line in mutations_and_primers:
		parts = line.strip('\n').split('\t')
		position = parts[6]
		if position == '' or position == 'Position':
			pass
		else:
			#print 'position =', position
			locations.append(position)

# list all files in app_results, theseare for the validated primers
validated_bespoke_primers = os.listdir('app_results')

# pandas df mutation, No. app primers, app primers, validated primer, match

mutations = list()
for primer_file in validated_bespoke_primers:
	# get the chromosome position of the mutation and add to a list or pandas df
	mutation = primer_file.split('.')[0]
	mutations.append(mutation)

bespoke_mutations = pd.DataFrame(mutations,columns=['Mutation'])
# open the file and count the number of primers generated by the app
for i in range(len(bespoke_mutations)):
	# pri_file = open(os.path.join(os.getcwd(),'app_results',bespoke_mutations['Mutations'][i]+'.csv'))
	num_primers = 0
	with open(os.path.join('app_results',bespoke_mutations['Mutation'][i]+'.csv'),'r') as pri_file:
		csv_file = csv.reader(pri_file)
		for row in csv_file:
			if row[0] == 'Gene':
				pass
			else:
				num_primers += 1
 
	# add number of primer pairs to df column called 'Primer3 primers (pairs)'
	bespoke_mutations.loc[i,'Primer3 primers (pairs)'] = int(num_primers)	